Code modernization and deduplication

zcobell opened the pull request:
# Description

This work consists of a series of commits which simplify and modernize the code. 

## Unification of Terminate Routines
Throughout the code there were many ways to terminate. This stemmed from the main termination routine being in `adcirc_mod` inside `adcirc.F`. Since this was there, it could easily create circular dependencies. The solution for some time was to place a copy of this routine inside different modules and let that module call its own terminate routine instead of introducing issues with circular dependencies. 

This work introduces a `mod_terminate` which only depends on `messenger`, and can then effectively be called anywhere in the code. This also removes all non-trivial calls to `exit` in favor of using the terminate routine. This should ensure that MPI is always correctly shut down, which I suspect was the source of errors where the code hung. 

Removing this duplicated logic eliminates ~1000 lines of code from the source with approximately 15 occurrences of the duplicated terminate routine being removed.

## Remove legacy syntax

There were portions of the code that used outdated syntax which can be safely removed and replaced with modern constructs (i.e. labeled do loops, etc) with the exception of `itpack.F` since it is a third-party module.

## Refactor Logging
The logging functions of the code were previously defined in `global.F`, however, these can be easily refactored into their own module which allows removal of dependencies on a global module. 

Additionally, the logging routines had a configuration that is set up to produce an (approximate) stack trace. The configuration requires that the user annotates their routines manually so that the code can know its (approximate) position in the call stack and echo it as necessary. To do this, two routines were configured:

```Fortran
call setMessageSource("my_function_name")

call unsetMessageSource()
```
It is required that these calls are paired in order to (a) maintain an accurate representation of the call stack but also (b) to ensure that the code doesn't walk off of either end of the array that is being used to track these variables. Because many functions in ADCIRC can have multiple return paths, this has implications on the user to ensure that all return paths are annotated. This can get rather messy and is easy to forget.

This work removes the need for this boilerplate and replaces it with a version of the code that forces the compiler to generate these log calls based on the lifetime of a variable which the compiler can track accurately. Instead, now the user will annotate the function with a single macro. This macro is REQUIRED to be the first line of code AFTER the variable declarations. The only other change is in the start of the source file to include the logging header which defines the macros appropriately (`#include "logging_macros.h"`) The code transformation (before/after) looks like this:

### Before
```Fortran
call setMessageSource("my_function")
#ifdef TRACE_THIS_MODULE
call logMessage("Enter")
#endif

... code ...

if(early_exit_condition)then
#ifdef TRACE_THIS_MODULE
    call logMessage("Exit")
#endif
    call unsetMessageSource()
    return
endif

... code ...

#ifdef TRACE_THIS_MODULE
call logMessage("Exit")
#endif
call unsetMessageSource()
    
```

### After
```Fortran
LOG_SCOPE_TRACED("my_function", TRACE_THIS_MODULE)
```

By using this compiler-generated logging infrastructure, we can eliminate an entire class of bugs that has plagued the codebase for some time.

## CMake updates
Using modern dependency tracking in cmake which helps with parallel builds and using alias targets to simplify cmake files.

## Help arguments
This commit adds help screens that can be passed to executables using `-h` or `--help`

### ADCIRC
```
Usage: adcirc [OPTION]...
ADvanced CIRCulation model for shelves, coastal seas, and estuaries.

General options:
  -h, --help             display this help and exit
  -v, --version          output version information and exit
      --hash             output git commit hash and exit

Report bugs to <https://github.com/adcirc/adcirc/issues>.
ADCIRC home page: <https://adcirc.org>.
User documentation: <https://adcirc.github.io/adcirc/>.
```

### PADCIRC
```
Usage: padcirc [OPTION]...
ADvanced CIRCulation model for shelves, coastal seas, and estuaries.

General options:
  -h, --help             display this help and exit
  -v, --version          output version information and exit
      --hash             output git commit hash and exit

Parallel execution options:
  -W NUM                 number of dedicated writer processors for output files
  -Ws NUM                number of dedicated writer processors with split file output
  -H NUM                 number of dedicated writer processors for hotstart files

Report bugs to <https://github.com/adcirc/adcirc/issues>.
ADCIRC home page: <https://adcirc.org>.
User documentation: <https://adcirc.github.io/adcirc/>.
```

### ADCPREP
```
Usage: adcprep [OPTION]...
Decompose ADCIRC input files for parallel execution.

Without command-line options, adcprep runs in interactive mode.

General options:
  -h, --help             display this help and exit
  -v, --version          output version information and exit
      --hash             output git commit hash and exit

Required option for all operations:
      --np NPROC         number of processors for parallel run

Operations (select one):
      --partmesh         partition mesh using METIS (nodal graph)
      --partmesh_npart_from_epart
                         partition nodes using element partition
      --prepall          decompose all input files
      --prep13           decompose nodal attributes file (fort.13) only
      --prep15           decompose run info file (fort.15) only
      --prep20           decompose time-varying flux file (fort.20) only

Mesh partitioning options:
      --strict-boundaries
                         enable strict boundary checking; retry partition
                         with modified weights if issues detected

File format options:
  -MFT, --mft TYPE       mesh file type: ascii or xdmf (default: ascii)
  -MFN, --mfn FILENAME   mesh filename (default: fort.14)
  -NFT, --nft TYPE       nodal attributes file type: ascii or xdmf
                         (default: ascii)
  -NFN, --nfn FILENAME
                         nodal attributes filename (default: fort.13)

Typical workflow:
  1. Partition the mesh:
       adcprep --np 256 --partmesh

  2. Decompose all input files:
       adcprep --np 256 --prepall

Report bugs to <https://github.com/adcirc/adcirc/issues>.
ADCIRC home page: <https://adcirc.org>.
User documentation: <https://adcirc.github.io/adcirc/>.
```

### ASWIP
```
Usage: aswip [OPTION]...
Process asymmetric hurricane track files and generate
input files for ADCIRC's parametric vortex models.

General options:
  -h, --help             display this help and exit
      --version          output version information and exit
      --hash             output git commit hash and exit

Wind model options:
  -N, -n NWS             NWS wind model type (19 or 20)
                         default: 20.
  -M, -m METHOD          method for selecting quadrant Rmax values
                         default: 4 (largest available isotach)
  -Z, -z APPROACH        approach for calculating quadrant Vr
                         default: 2 (tangential to azimuth)
  -G, -g FACTOR          geostrophic (1) or cyclostrophic (0) balance
                         default: 1

Input/output file options:
  -W, -w FILENAME        track file to read (default: fort.22)
  -I, -i FILENAME        fort.14 mesh file for spatial output
  -L, -l FILENAME        text file with station locations
  -E, -e FILENAME        element table file for spatial output
  -D, -d                 do not write NWS19 fort.22 output

Parameter adjustment options:
  -R, -r RMAX            constant forecast Rmax in nautical miles
                         default: calculated from isotach data
  -P, -p PERCENT         percent to add to calculated Rmax
                         (e.g., 10 = 10% increase, -10 = 10% decrease)
                         default: no modification
  -X, -x PERCENT         percent to add to Vmax
                         default: 0 (no modification)
  -T, -t TIME            time offset in seconds (for hotstart)
                         default: 0

Diagnostic output options:
  -A, -a                 write full circle Rmax values (360 deg)
  -V, -v                 write radial velocity and pressure
  -S, -s                 write spatial V and P on mesh
  -B, -b                 write uvp at node locations
  -C, -c                 write spatial xy & uvp on mesh
  -F, -f FRAME           select particular frame to render
  -O, -o SECONDS         time increment for output

Examples:
  Process ATCF track file for NWS20:
    aswip -w best_track.dat -n 20

  Generate wind fields on mesh with spatial output:
    aswip -i fort.14 -s -n 20

Report bugs to <https://github.com/adcirc/adcirc/issues>.
ADCIRC home page: <https://adcirc.org>.
User documentation: <https://adcirc.github.io/adcirc/>.
```

## Removing obsolete compiler options/flags
There were different compiler options that were introduced throughout the codebase for older compilers. Some of these options were because older Fortran standards (pre-Fortran 90 or 2003) which did not have a well specified internal read/write, and therefore ADCIRC carried its own implementations of these, which were unlocked using a combination of `-DLINUX` and a machine specific flag. 

Others were specific to certain systems or compilers (i.e. `-DIBM`, `-DVEC`, `-DMACHSUN`, `-DSCA`, `-DCRAY`, `-DCRAYX1`, etc.) which are no longer used (the last vector computer has long since been decommissioned and auto-vectorizing compilers are the norm). 

Additionally, the compiler flag `-mcmodel=medium` was often included, but there are no large static arrays in the code and therefore it is not necessary. 

## Cleaning up cmake build system
The cmake build system moves to a more modern `target_link_libraries` syntax to specify dependencies rather than `add_dependencies` and interface targets/aliases are used for include directories and library linkages. macros are replaced with functions so that variables are locally scoped. 

## Using `RPATH` linkage to avoid missing/mismatched libraries at runtime
CMake will now by default pass `RPATH` arguments to the linker to ensure the libraries used at compile time are used at runtime and is generally safer rather than relying on `LD_LIBRARY_PATH` to be set correctly. This can optionally be disabled on user request.

## Type of change

<!--- Select the type of change, use [x] to select and [ ] to mark unselected -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Bug fix with breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to existing feature to add new functionality
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

<!--- Please fill out the checklist. Use [x] to mark selected and [ ] to mark unselected -->

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] My code is up-to-date and tested with changes from the latest version of `main`

zcobell committed:
Unifying terminate routines

There were about 15 different implementations of the
terminate routine. This unifies them by placing them
in a single, common, module so it does not introduce
dependency issues

Removing calls to "exit"

zcobell committed:
Removing legacy fortran subroutines and syntax

1. Older versions of fortran required machine
specific subroutines to handle things now covered
in the Fortran standard. Replacing the old versions
with their modern equivilents

2. Modernizing do-loop syntax and variable declarations

zcobell committed:
Adding a help info screen to the command line for executables

zcobell committed:
Modernizing and cleaning up cmake build system

1. Removing obsolete compiler options

Old Cray, IBM, and SUN flags are removed.

Old flags differentiating between scalar and vector
computing systems are removed since modern compilers are
all automatically vectorizing. The last vector computer
(Cray X1) was released in 2003 and decomissioned in 2010

2. Updating architecture checks

Removing mcmodel flags since they are not used

3. Adding RPATH linkage to cmake builds

This helps avoid issues with not finding libraries at runtime

4. Modernizing cmake macros

The cmake macros are converted to functions and modern linkage
techniques are used to better build the dependency graph

zcobell committed:
Fixing minor bug in sponge layer spherical coordinates

zcobell committed:
Refactoring logging routines into their own module

Updating so that nabout is of type t_log_level

acrosby commented:
Is cmake 3.14+ really required or can the requirement be changed to 3.10+ or something?

Edit: Seems like it actually might be.

zcobell commented:
Yea, I think 3.10 was missing some of the newer features. It's pretty old actually (~2017 or so).

I did a quick dig onto versions that are available in LTS operating system distributions and 3.17 is the minimum that is currently in the wild for any OS that is under non-specialized vendor support.

zcobell committed:
Using RAII based logger to set/unset the message sources

ADCIRC followed the pattern to generate a call stack using
a call to setMessageSource at the start of a function and
a call to unsetMessageSource when it left the function. This
is fine, however, a mismatched pair of calls could cause the
code to walk off the front or back of the messageSources array.

This means that the user needs to correctly annotate every entry
and return path for a function or risk a bug. What this work does
is employ the compiler to handle these tasks. The user specifies
a call to a macro:

LOG_SCOPE_TRACED("myfunction", trace_condition)

This automatically instantiates a variable with type t_log_scope
and makes the calls to setMessageSource. On its own this isn't a
huge improvement (apart from the fact that the boilerplate of
calling "call allMessage("Enter")" is removed), however the nice
part about this macro is it will call the destructor when the
variable goes out of scope. This means that there is no need to
annotate return paths and the compiler will handle this correctly

This has been tested with gcc 8 and above, Intel compilers, and
nvhpc compilers. Compilers adhering to the Fortran 2008 standard
(aside from coarray Fortran) will handle this correctly.

zcobell committed:
Cleaning up unused MPI error return codes

zcobell committed:
Adding LTO option to cmake

zcobell committed:
Fixing Intel variable initializations
